tensorflow/python/keras/layers/preprocessing/preprocessing_stage_functional.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Preprocessing stage."""
+# pylint: disable=g-classes-have-attributes
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+
+from tensorflow.python.data.ops import dataset_ops
+from tensorflow.python.framework import ops
+from tensorflow.python.keras.engine import base_preprocessing_layer
+from tensorflow.python.keras.engine import functional
+from tensorflow.python.keras.utils import tf_utils
+from tensorflow.python.util import nest
+
+
+class FunctionalPreprocessingStage(base_preprocessing_layer.PreprocessingLayer,
+                                   functional.Functional):
+  """A functional preprocessing stage.
+
+  This preprocessing stage wraps a graph of preprocessing layers into a
+  Functional-like object that enables you to `adapt()` the whole graph via
+  a single `adapt()` call on the preprocessing stage.
+
+  Arguments:
+    inputs: List of input tensors (must be created via `tf.keras.Input()`).
+    outputs: List of outputs tensors.
+    name: String, optional. Name of the model.
+    trainable: Boolean, whether the model's variables should be trainable.
+  """
+
+  def fit(self, *args, **kwargs):
+    raise ValueError(
+        'Preprocessing stage is not a complete model, and hence should not be '
+        '`fit`. Instead, you may feed data to `adapt` the stage to set '
+        'appropriate states of the layers in the stage.')
+
+  def adapt(self, data, reset_state=True):
```

Comment:
Does this allow a string-keyed dict of tensors? What about as a Dataset structure?

Reply:
yes it allows.

Reply:
Can we add that to the documentation?

Reply:
I added to the docstring. Can you take a look if I missed anything? Thanks!